client/client.go

```diff
@@ -68,56 +73,62 @@ func New(messageService messageservice.MessageService, chainservice chainservice
 	c.receivedVouchers = make(chan payments.Voucher, 1000)
 
 	c.channelNotifier = notifier.NewChannelNotifier(store, c.vm)
+
 	// Start the engine in a go routine
+
 	go c.engine.Run()
 
+	c.wg = &sync.WaitGroup{}
+	c.wg.Add(1)
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancelEventHandler = cancel
 	// Start the event handler in a go routine
 	// It will listen for events from the engine and dispatch events to client channels
-	go c.handleEngineEvents()
+	go c.handleEngineEvents(ctx)
 
 	return c
 }
 
 // handleEngineEvents is responsible for monitoring the ToApi channel on the engine.
 // It parses events from the ToApi chan and then dispatches events to the necessary client chan.
-func (c *Client) handleEngineEvents() {
-	for update := range c.engine.ToApi() {
-
-		for _, completed := range update.CompletedObjectives {
-			d, _ := c.completedObjectives.LoadOrStore(string(completed.Id()), make(chan struct{}))
-			close(d)
-
-			// use a nonblocking send to the RPC Client in case no one is listening
-			select {
-			case c.completedObjectivesForRPC <- completed.Id():
-			default:
+func (c *Client) handleEngineEvents(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
We actually don't have to use a `context` to signal `handleEngineEvents` to stop executing, we could just rely on the engine `ToAPI` chan getting closed when  we call `engine.Close` like we were doing previously.  

However I think this is less efficient as `close` has ["the effect of shutting down the channel after the last sent value is received"](https://pkg.go.dev/builtin#close), meaning this goroutine will only stop executing after it's gotten through the buffer of the `chan`.

By using a `context` and `cancel` the goroutine will stop as soon as it reads from `ctx.Done`

Reply:
I suppose the select statement could still end up "ignoring" the `ctx.Done()` path for some time as the buffer empties? It seems more likely that the goroutine is stopped sooner than via the existing behaviour, but I guess no certainty about that?